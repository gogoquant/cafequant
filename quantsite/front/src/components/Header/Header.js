import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Layout, Menu, Avatar, Icon } from 'antd';
import classNames from 'classnames';
import styles from './Header.scss';

const menuConfig = [
  {
    key: 'all',
    name: '全部',
  },
  {
    key: 'good',
    name: '精华',
  },
  {
    key: 'share',
    name: '分享',
  },
  {
    key: 'ask',
    name: '问答',
  },
  {
    key: 'job',
    name: '招聘',
  },
  // {
  //   key: 'dev',
  //   name: '测试',
  // },
];

const headerCls = classNames(styles.header, styles.lgContainer);

const User = ({ name, avatar }) => (
  <span>
    <Avatar className={styles.avatar} size="small" src={avatar} />
    {name}
  </span>
);

const Account = ({ isLogin, name, avatar, onClickMenu }) => {
  if (isLogin) {
    return (
      <div className={styles.right}>
        <Menu mode="horizontal" selectable={false} onClick={onClickMenu}>
          <Menu.SubMenu
            className={styles.account}
            title={<User name={name} avatar={avatar} />}
          >
            <Menu.Item key="user">
              <Icon type="user" />用户中心
            </Menu.Item>
            <Menu.Divider />
            <Menu.Item key="logout">
              <Icon type="logout" />退出登录
            </Menu.Item>
          </Menu.SubMenu>
        </Menu>
      </div>
    );
  }
  return (
    <div className={styles.right}>
      <Menu
        className={styles.account}
        mode="horizontal"
        selectable={false}
        onClick={onClickMenu}
      >
        <Menu.Item key="login">登录</Menu.Item>
      </Menu>
    </div>
  );
};

class Header extends React.Component {
  shouldComponentUpdate(nextProps) {
    return (
      nextProps.isMobile !== this.props.isMobile ||
      nextProps.user !== this.props.user
    );
  }

  render() {
    const { isMobile, user, onClickMenu } = this.props;
    return (
      <Layout.Header className={styles.container}>
        <div className={headerCls}>
          <Link to="/">
            <div className={styles.logo} />
          </Link>
          <Menu
            className={styles.menu}
            mode="horizontal"
            defaultSelectedKeys={['all']}
            onClick={onClickMenu}
          >
            {menuConfig.map(item => (
              <Menu.Item key={item.key}>{item.name}</Menu.Item>
            ))}
          </Menu>
          {!isMobile && (
            <Account
              isLogin={user.id !== undefined}
              name={user.name}
              avatar={user.avatar}
              onClickMenu={onClickMenu}
            />
          )}
        </div>
      </Layout.Header>
    );
  }
}

User.propTypes = {
  name: PropTypes.string,
  avatar: PropTypes.string,
};

Account.propTypes = {
  isLogin: PropTypes.bool.isRequired,
  name: PropTypes.string,
  avatar: PropTypes.string,
  onClickMenu: PropTypes.func.isRequired,
};

Header.propTypes = {
  isMobile: PropTypes.bool.isRequired,
  user: PropTypes.object.isRequired,
  onClickMenu: PropTypes.func.isRequired,
};

export default Header;
