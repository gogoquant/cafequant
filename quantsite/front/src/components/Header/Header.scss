@import "../../styles/Base";

.container {
  width: 100%;
  background: white;
  height: auto;
  display: flex;
  justify-content: center;
  box-shadow: 0 1px 4px rgba(0, 21, 41, .08);
}

.header {
  height: 100%;
  display: flex;
  align-items: center;
  @media screen and(max-width: $screen-md) {
    flex-direction: column;
  }
}

.logo {
  width: 128px;
  height: 32px;
  float: left;
  margin: 16px 24px;
  background: url("../../assets/logo.svg");
}

.menu {
  width: 100%;
  line-height: 64px;
  @media screen and(max-width: $screen-md) {
    display: flex;
    justify-content: center;
    flex-wrap: wrap;
  }
}

.right {
  flex: 1;
  display: flex;
  justify-content: flex-end;
}

.account {
  line-height: 64px;
}

.avatar {
  margin-right: 8px;
}
