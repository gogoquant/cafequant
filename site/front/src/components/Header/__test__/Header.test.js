import React, { cloneElement } from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import Header from '../Header';

describe('<Header />', () => {
  const props = {
    isMobile: false,
    user: {},
    onClickMenu: jest.fn(),
  };
  const tree = mount(
    <MemoryRouter>
      <Header {...props} />
    </MemoryRouter>,
  );

  it('should render login', () => {
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('should re-render when isMobile or user change', () => {
    tree.setProps({
      children: cloneElement(tree.props().children, {
        user: { id: '007', name: 'admin' },
      }),
    });
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });
});
