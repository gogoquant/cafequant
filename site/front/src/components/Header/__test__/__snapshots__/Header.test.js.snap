// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should re-render when isMobile or user change 1`] = `
<header
  className="ant-layout-header"
>
  <div
    className=""
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <div />
    </a>
    <ul
      className="ant-menu ant-menu-light ant-menu-root ant-menu-horizontal"
      onMouseEnter={[Function]}
      onTransitionEnd={[Function]}
      role="menu"
      style={Object {}}
    >
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item ant-menu-item-selected"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        全部
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        精华
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        分享
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        问答
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        招聘
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "position": "absolute",
            "visibility": "hidden",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
    </ul>
    <div>
      <ul
        className="ant-menu ant-menu-light ant-menu-root ant-menu-horizontal"
        onMouseEnter={[Function]}
        onTransitionEnd={[Function]}
        role="menu"
        style={Object {}}
      >
        <li
          className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-menu-submenu-title"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={Object {}}
          >
            <span>
              ···
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          className="ant-menu-submenu ant-menu-submenu-horizontal"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
        >
          <div
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-menu-submenu-title"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={Object {}}
          >
            <span>
              <span
                className="ant-avatar ant-avatar-sm ant-avatar-circle"
                style={Object {}}
              >
                <span
                  className="ant-avatar-string"
                />
              </span>
              admin
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
          style={
            Object {
              "position": "absolute",
              "visibility": "hidden",
            }
          }
        >
          <div
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-menu-submenu-title"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={Object {}}
          >
            <span>
              ···
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
    </div>
  </div>
</header>
`;

exports[`<Header /> should render login 1`] = `
<header
  className="ant-layout-header"
>
  <div
    className=""
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <div />
    </a>
    <ul
      className="ant-menu ant-menu-light ant-menu-root ant-menu-horizontal"
      onMouseEnter={[Function]}
      onTransitionEnd={[Function]}
      role="menu"
      style={Object {}}
    >
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item ant-menu-item-selected"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        全部
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        精华
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        分享
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        问答
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "display": "none",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        className="ant-menu-item"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={Object {}}
      >
        招聘
      </li>
      <li
        className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="menuitem"
        style={
          Object {
            "position": "absolute",
            "visibility": "hidden",
          }
        }
      >
        <div
          aria-expanded={false}
          aria-haspopup="true"
          className="ant-menu-submenu-title"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={Object {}}
        >
          <span>
            ···
          </span>
          <i
            className="ant-menu-submenu-arrow"
          />
        </div>
      </li>
    </ul>
    <div>
      <ul
        className="ant-menu ant-menu-light ant-menu-root ant-menu-horizontal"
        onMouseEnter={[Function]}
        onTransitionEnd={[Function]}
        role="menu"
        style={Object {}}
      >
        <li
          className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-menu-submenu-title"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={Object {}}
          >
            <span>
              ···
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          className="ant-menu-item"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
          style={Object {}}
        >
          登录
        </li>
        <li
          className="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          role="menuitem"
          style={
            Object {
              "position": "absolute",
              "visibility": "hidden",
            }
          }
        >
          <div
            aria-expanded={false}
            aria-haspopup="true"
            className="ant-menu-submenu-title"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={Object {}}
          >
            <span>
              ···
            </span>
            <i
              className="ant-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
    </div>
  </div>
</header>
`;
